accord-core/src/main/java/accord/coordinate/Barrier.java

```diff
@@ -0,0 +1,301 @@
+            {
+                Timestamp executeAt = command.executeAt();
+                // In all the cases where we add a listener (listening to existing command, async completion of CoordinateSyncPoint)
+                // we want to notify the agent
+                doBarrierSuccess(executeAt);
+            }
+            else if (command.is(Status.Invalidated))
+                // TODO is this the right way to error out?
+                Barrier.this.tryFailure(new Timeout(command.txnId(), command.homeKey()));
+
+        }
+
+        @Override
+        public PreLoadContext listenerPreLoadContext(TxnId caller)
+        {
+            // TODO Surely the command passed to `onChange` is loaded?
+            return EMPTY_PRELOADCONTEXT;
+        }
+
+        @Override
+        public boolean isTransient()
+        {
+            return true;
+        }
+    }
+
+    private ExistingTransactionCheck checkForExistingTransaction()
+    {
+        // TODO this could execute at several different command stores and register multiple command listeners
+        // Won't harm correctness, but maybe wasteful?
+        ExistingTransactionCheck check = new ExistingTransactionCheck();
+        node.commandStores().mapReduceConsume(
+                contextFor(keyOrRange.asKey()),
+                keyOrRange.asKey().toUnseekable(),
+                minEpoch,
+                Long.MAX_VALUE,
+                check);
+        return check;
+    }
+
+    // Hold result of looking for a transaction to act as a barrier for an Epoch
+    static class BarrierTxn
+    {
+        @Nonnull
+        public final TxnId txnId;
+        @Nonnull
+        public final Timestamp executeAt;
+        @Nonnull
+        public final Status status;
+        public BarrierTxn(@Nonnull TxnId txnId, @Nonnull Timestamp executeAt, @Nonnull Status status)
+        {
+            this.txnId = txnId;
+            this.executeAt = executeAt;
+            this.status = status;
+        }
+
+        public BarrierTxn max(BarrierTxn other)
+        {
+            if (other == null)
+                return this;
+            return status.compareTo(other.status) >= 0 ? this : other;
+        }
+    }
+
+    /*
+     * Check for an existing transaction that is either already Applied (dependencies were applied)
+     * or PreApplied (outcome fixed, dependencies not yet applied).
+     *
+     * For Applied we can return success immediately with the executeAt epoch. For PreApplied we can add
+     * a listener for when it transitions to Applied and then return success.
+     */
+    class ExistingTransactionCheck extends AsyncResults.SettableResult<BarrierTxn> implements MapReduceConsume<SafeCommandStore, BarrierTxn>
+    {
+        @Override
+        public BarrierTxn apply(SafeCommandStore safeStore)
+        {
+            // TODO Is this going to find the correct command store to listen for transaction progress?
+            BarrierTxn found = safeStore.mapReduceWithTerminate(
```

Comment:
Let's leave this unresolved for now. I removed some of them, but I still want to go over it with Benedict and I am pretty sure that `TestKind.Any` doesn't work because an exclusive sync doesn't wait for dependencies to apply.